contrib/epee/include/span_view.h

```diff
@@ -0,0 +1,174 @@
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include <cstddef>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+namespace epee
+{
+  /*!
+    \brief Non-owning sequence of data. Does not deep copy
+
+    Inspired by `gsl::span`. This class is intended to be used as a parameter
+    type for functions that need to take a writable or read-only sequence of
+    data. Most common cases are `span<char>` and `span<std::uint8_t>`. Using as
+    a class member is only recommended if clearly documented as not doing a
+    deep-copy. `std::vector`, `std::string`, `std::array`, and C-arrays are all
+    easily convertible to this type.
+
+    \note Conversion from C string literal to `view<char>` will include the
+      NULL-terminator.
+    \note `view<T>` is an alias for `span<const T>` which is read-only.
+    \note Never allows derived-to-base pointer conversion; an array of derived
+      types is not an array of base types.
+   */
+  template<typename T>
+  class span
+  {
+    /* Supporting class types is tricky - the {ptr,len} and {container}
+       constructors will allow derived-to-base conversions. This is NOT
+       desireable because an array of derived types is not an array of base
+       types. It is possible to handle this case, implement when/if needed. */
+    static_assert(std::is_integral<T>::value, "only integral values currently for span");
+
+    // Disables ADL for these calls.
+    struct get {
+      // we only accept pointer and size_t return types, so use noexcept aggressively
+      template<std::size_t N>
+      static constexpr T* data(T (&src)[N]) noexcept { return src; }
+
+      template<typename C>
+      static constexpr auto
+      data(C& src) noexcept(noexcept(src.data())) -> decltype(src.data()) {
+        return src.data();
+      }
+
+
+      template<std::size_t N>
+      static constexpr std::size_t size(T (&)[N]) noexcept { return N; }
+
+      template<typename C>
+      static constexpr auto
+      size(const C& src) noexcept(noexcept(src.size())) -> decltype(src.size()) {
+        return src.size();
+      }
+    };
+
+  public:
+    using value_type = T;
+    using pointer = T*;
+    using reference = T&;
+    using iterator = pointer;
+    using const_iterator = typename std::add_const<value_type>::type *;
+
+    constexpr span() noexcept : ptr(nullptr), len(0) {}
+    constexpr span(std::nullptr_t) noexcept : span() {}
```

Comment:
Is that calling a ctor from a ctor ? I always wanted to do this, but AFAIK it was not allowed. Is this a new change in C++ ?